File: built-in plugins dsp/mb-mix-vca-group.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbMixVcaGroupPlugin: PluginDefinition = { id: 'mb-mix-vca-group', slug: 'mb-mix-vca-group', name: 'MB VCA Group', category: 'effect', type: 'mixing' as any, version: '1.0.0', description: 'VCA-style group fader with smooth gain control', author: 'Max Booster', parameters: [{ id: 'gain', name: 'Gain', type: 'float', defaultValue: 0, minValue: -60, maxValue: 12, automatable: true }, { id: 'trim', name: 'Trim', type: 'float', defaultValue: 0, minValue: -12, maxValue: 12, automatable: true }, { id: 'link', name: 'Stereo Link', type: 'float', defaultValue: 1, minValue: 0, maxValue: 1, automatable: false }], defaultPreset: { gain: 0, trim: 0, link: 1 } };

export default MbMixVcaGroupPlugin;

File: built-in plugins dsp/mb-mix-vca-group.cpp
/**
 * MB VCA Group
 * Category : effect
 * Type     : mixing
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : VCA-style group fader with smooth gain control
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_MIX_VCA_GROUP_H
#define MB_MIX_VCA_GROUP_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbMixVcaGroup : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-mix-vca-group";
    static constexpr const char* PLUGIN_NAME    = "MB VCA Group";
    static constexpr const char* PLUGIN_TYPE    = "mixing";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float gain = 0f;  // range [-60, 12]
    float trim = 0f;  // range [-12, 12]
    float link = 1f;  // range [0, 1]
    };

    MbMixVcaGroup() = default;
    ~MbMixVcaGroup() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.gain = std::clamp(params.gain, -60f, 12f);
        params.trim = std::clamp(params.trim, -12f, 12f);
        params.link = std::clamp(params.link, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB VCA Group
        return input;
    }
};

#endif // MB_MIX_VCA_GROUP_H
